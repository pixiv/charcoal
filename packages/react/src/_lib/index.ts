/**
 * 今後ポートされる予定の汎用的な関数群
 */

/**
 * Function used to assert a given code path is unreachable
 */
export function unreachable(): never
/**
 * Function used to assert a given code path is unreachable.
 * Very useful for ensuring switches are exhaustive:
 *
 * ```ts
 * switch (a.type) {
 *   case Types.A:
 *   case Types.B:
 *     break
 *   default:
 *     unreachable(a) // will cause a build error if there was
 *                    // a Types.C that was not checked
 * }
 * ```
 *
 * @param value Value to be asserted as unreachable
 */
// NOTE: Uses separate overloads, _not_ `value?: never`, to not allow `undefined` to be passed

export function unreachable(value: never): never
export function unreachable(value?: never): never {
  throw new Error(
    arguments.length === 0
      ? 'unreachable'
      : `unreachable (${JSON.stringify(value)})`
  )
}

/**
 * 複数のrefをマージする。
 *
 * forwardRefで受け取ったrefと、コンポーネント内で定義したrefを同じ要素につけたいケースなどで使う
 */
export function mergeRefs<T>(...refs: React.Ref<T>[]): React.RefCallback<T> {
  return (value) => {
    for (const ref of refs) {
      if (typeof ref === 'function') {
        ref(value)
      } else if (ref !== null) {
        // TODO: use RefObject
        // https://react.dev/blog/2024/04/25/react-19-upgrade-guide#:~:text=%3D%201%3B-,MutableRef,-is%20now%20deprecated
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    }
  }
}

export function countCodePointsInString(string: string): number {
  // [...string] とするとproduction buildで動かなくなる
  // cf. https://twitter.com/f_subal/status/1497214727511891972
  return Array.from(string).length
}
