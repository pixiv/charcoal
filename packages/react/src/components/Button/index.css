.charcoal-button {
  appearance: none;
  background: transparent;
  box-sizing: border-box;
  padding: 0 24px;
  border-style: none;
  outline: none;
  text-rendering: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
  width: min-content;
  display: inline-grid;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  user-select: none;
  white-space: nowrap;
  border-radius: 999999px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  color: var(--charcoal-color-text-default);
  background-color: var(--charcoal-color-container-secondary-default-a);
  transition: 0.2s color, 0.2s background-color, 0.2s box-shadow;
  height: 40px;
}

.charcoal-button:disabled {
  cursor: default;
  opacity: 0.32;
}

.charcoal-button:not(:disabled):focus-visible {
  outline: none;
  box-shadow: var(--charcoal-color-custom-focus-effect);
}

.charcoal-button:not(:disabled):hover {
  color: var(--charcoal-color-text-hover);
  background-color: var(--charcoal-color-container-secondary-hover-a);
}

.charcoal-button:not(:disabled):active,
.charcoal-button:not(:disabled)[data-active='true'] {
  color: var(--charcoal-color-text-press);
  background-color: var(--charcoal-color-container-secondary-press-a);
}

.charcoal-button[data-variant='Primary'] {
  color: var(--charcoal-color-text-on-primary-default);
  background-color: var(--charcoal-color-container-primary-default);
}

.charcoal-button[data-variant='Primary']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-primary-hover);
  background-color: var(--charcoal-color-container-primary-hover);
}

.charcoal-button[data-variant='Primary']:active:not(:disabled),
.charcoal-button[data-variant='Primary'][data-active='true'] {
  color: var(--charcoal-color-text-on-primary-press);
  background-color: var(--charcoal-color-container-primary-press);
}

.charcoal-button[data-variant='Overlay'] {
  color: var(--charcoal-color-text-on-on-img-default);
  background-color: var(--charcoal-color-container-on-img-default);
}

.charcoal-button[data-variant='Overlay']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-on-img-hover);
  background-color: var(--charcoal-color-container-on-img-hover);
}

.charcoal-button[data-variant='Overlay']:active:not(:disabled),
.charcoal-button[data-variant='Overlay'][data-active='true'] {
  color: var(--charcoal-color-text-on-on-img-press);
  background-color: var(--charcoal-color-container-on-img-press);
}

/*
 * HUD 要素ではないが、以下の理由で HUD を色として利用する。
 * 1. UI パッケージであるがゆえにデザインを大幅に変えたくない
 *   a. このコンポーネントデザインは改善できると思っている
 * 2. この要素を説明するための Applied Token が存在しないため、例外的に Component Token という位置づけを追加することになる
 * 改善するとなくなる要素のために新しく Token を追加することは避けたいので、今回は最も近しい Applied Token の HUD を利用することにした。
 */
.charcoal-button[data-variant='Navigation'] {
  color: var(--charcoal-color-text-on-hud-default);
  background-color: var(--charcoal-color-container-hud-default);
}

.charcoal-button[data-variant='Navigation']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-primary-hover);
  background-color: var(--charcoal-color-container-hud-hover);
}

.charcoal-button[data-variant='Navigation']:active:not(:disabled),
.charcoal-button[data-variant='Navigation'][data-active='true'] {
  color: var(--charcoal-color-text-on-primary-press);
  background-color: var(--charcoal-color-container-hud-press);
}

.charcoal-button[data-variant='Danger'] {
  color: var(--charcoal-color-text-on-negative-default);
  background-color: var(--charcoal-color-container-negative-default);
}
.charcoal-button[data-variant='Danger']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-negative-hover);
  background-color: var(--charcoal-color-container-negative-hover);
}

.charcoal-button[data-variant='Danger']:active:not(:disabled),
.charcoal-button[data-variant='Danger'][data-active='true'] {
  color: var(--charcoal-color-text-on-negative-press);
  background-color: var(--charcoal-color-container-negative-press);
}

.charcoal-button[data-size='S'] {
  padding: 0 16px;
  height: 32px;
}

.charcoal-button[data-full-width='true'] {
  width: 100%;
}
