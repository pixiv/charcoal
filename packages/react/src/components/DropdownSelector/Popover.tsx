import { FocusScope } from '@react-aria/focus'
import { DismissButton, useOverlay } from '@react-aria/overlays'
import React, {
  FC,
  useRef,
  useMemo,
  PropsWithChildren,
  memo,
  CSSProperties,
} from 'react'
import { mergeProps } from '@react-aria/utils'

type Props = PropsWithChildren<{
  open?: boolean
  onClose?: () => void
  style?: CSSProperties
  className?: string
}>

const Popover: FC<Props> = ({ open, onClose, children, ...props }) => {
  const ref = useRef<HTMLDivElement>(null)

  const { overlayProps } = useOverlay(
    useMemo(
      () => ({
        isOpen: open,
        onClose,
        shouldCloseOnBlur: true,
        isDismissable: true,
      }),
      [onClose, open]
    ),
    ref
  )

  return (
    <FocusScope restoreFocus>
      <div {...mergeProps(overlayProps, props)} ref={ref}>
        {children}
        <DismissButton onDismiss={onClose} />
      </div>
    </FocusScope>
  )
}

export default memo(Popover)
