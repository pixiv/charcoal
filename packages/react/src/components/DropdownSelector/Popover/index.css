.charcoal-popover {
  margin: 4px 0;
  list-style: none;
  overflow: auto;
  max-height: inherit;
  background-color: var(--charcoal-color-background-default);
  border: solid 1px var(--charcoal-color-border-secondary);
  border-radius: 8px;
  padding-top: 8px;
  padding-bottom: 8px;
}
