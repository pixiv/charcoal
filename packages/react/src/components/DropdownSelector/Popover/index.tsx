import React, { RefObject, useRef } from 'react'
import { ReactNode } from 'react'
import { DismissButton, Overlay, usePopover } from '@react-aria/overlays'
import styled from 'styled-components'
import { theme } from '../../../styled'

export type PopoverProps = {
  isOpen: boolean
  onClose: () => void
  children: ReactNode
  triggerRef: RefObject<Element>
  popoverRef?: RefObject<HTMLDivElement>
}

const _empty = () => null

/**
 * 画面の全面に動的に開くことができるコンテナ要素
 * 外の要素をクリックしたり、内部からフォーカスを移動した場合に自動的に閉じる
 *
 * triggerRefの付近に画面内に収まるように表示される
 */
export default function Popover(props: PopoverProps) {
  const defaultPopoverRef = useRef<HTMLDivElement>(null)
  const finalPopoverRef =
    props.popoverRef === undefined ? defaultPopoverRef : props.popoverRef
  const { popoverProps, underlayProps } = usePopover(
    {
      triggerRef: props.triggerRef,
      popoverRef: finalPopoverRef,
      containerPadding: 16,
    },
    {
      close: props.onClose,
      isOpen: props.isOpen,
      // never used
      open: _empty,
      setOpen: _empty,
      toggle: _empty,
    }
  )

  if (!props.isOpen) return null

  return (
    <Overlay portalContainer={document.body}>
      <div {...underlayProps} style={{ position: 'fixed', inset: 0 }} />
      <DropdownPopoverDiv {...popoverProps} ref={finalPopoverRef}>
        <DismissButton onDismiss={() => props.onClose()} />
        {props.children}
        <DismissButton onDismiss={() => props.onClose()} />
      </DropdownPopoverDiv>
    </Overlay>
  )
}

const DropdownPopoverDiv = styled.div`
  margin: 4px 0;
  list-style: none;
  overflow: auto;
  max-height: inherit;

  ${theme((o) => [
    o.bg.background1,
    o.border.default,
    o.borderRadius(8),
    o.padding.vertical(8),
  ])}
`
