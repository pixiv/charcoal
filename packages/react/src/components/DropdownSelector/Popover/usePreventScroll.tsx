import { useEffect } from 'react'

export function usePreventScroll(
  element: HTMLElement | null,
  isOpen: boolean,
  useClip = false
): void {
  useEffect(() => {
    if (isOpen && element) {
      const defaultPaddingRight = element.style.paddingRight
      const defaultOverflow = element.style.overflow
      element.style.paddingRight = `${
        window.innerWidth - element.clientWidth
      }px`
      element.style.overflow = useClip ? 'clip' : 'hidden'
      return () => {
        element.style.paddingRight = defaultPaddingRight
        element.style.overflow = defaultOverflow
      }
    }
  }, [element, isOpen, useClip])
}
