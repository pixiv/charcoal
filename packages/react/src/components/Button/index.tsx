import React, { ReactNode, Ref } from 'react'
import styled from 'styled-components'
import { unreachable } from '../../_lib'
import { theme } from '../../styled'
import { Spacing, ThemeColor } from '@charcoal-ui/theme'

type Variant = 'Primary' | 'Default' | 'Overlay' | 'Danger' | 'Navigation'
type Size = 'S' | 'M'

type CustomJSXElement =
  | keyof JSX.IntrinsicElements
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | React.JSXElementConstructor<any>

export type ButtonProps<T extends CustomJSXElement = 'button'> = {
  /**
   * ボタンのスタイル
   */
  variant?: Variant
  /**
   * ボタンのサイズ
   */
  size?: Size
  /**
   * 幅を最大まで広げて描画
   */
  fullWidth?: boolean
  as?: T
  children?: ReactNode
} & Omit<React.ComponentProps<T>, 'children'>

function Button<T extends CustomJSXElement>(
  {
    children,
    variant = 'Default',
    size = 'M',
    fullWidth = false,
    ...rest
  }: ButtonProps<T>,
  ref: Ref<HTMLButtonElement>
) {
  const colors = variantToProps(variant)
  return (
    <StyledButton
      {...rest}
      $height={size === 'M' ? 40 : 32}
      $padding={size === 'M' ? 24 : 16}
      $font={colors.font}
      $background={colors.background}
      $fullWidth={fullWidth}
      ref={ref}
    >
      {children}
    </StyledButton>
  )
}
const ForwardedComponent = React.forwardRef<HTMLButtonElement, ButtonProps>(
  Button
) as typeof Button

export default ForwardedComponent

type StyledButtonProps = {
  $fullWidth: boolean
  $height: number
  $font: keyof ThemeColor
  $background: keyof ThemeColor
  $padding: keyof Spacing
}

const StyledButton = styled.button<StyledButtonProps>`
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;

  &:focus {
    outline: none;
  }

  /* Change the font styles in all browsers. */
  font: inherit;

  /* Remove the margin in Firefox and Safari. */
  margin: 0;

  /* Show the overflow in Edge. */
  overflow: visible;

  /* Remove the inheritance of text transform in Firefox. */
  text-transform: none;

  /* Remove the inner border and padding in Firefox. */
  &::-moz-focus-inner {
    border-style: none;
    padding: 0;
  }

  width: ${(p) => (p.$fullWidth ? 'stretch' : 'min-content')};
  display: inline-grid;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  user-select: none;
  white-space: nowrap;

  ${(p) =>
    theme((o) => [
      o.font[p.$font].hover.press,
      o.bg[p.$background].hover.press,
      o.typography(14).bold.preserveHalfLeading,
      o.padding.horizontal(p.$padding),
      o.disabled,
      o.borderRadius('oval'),
      o.outline.default.focus,
    ])}

  height: ${(p) => p.$height}px;
`

function variantToProps(variant: Variant) {
  switch (variant) {
    case 'Overlay':
      return { font: 'text5', background: 'surface4' } as const
    case 'Default':
      return { font: 'text2', background: 'surface3' } as const
    case 'Primary':
      return { font: 'text5', background: 'brand' } as const
    case 'Navigation':
      return { font: 'text5', background: 'surface6' } as const
    case 'Danger':
      return { font: 'text5', background: 'assertive' } as const
    default:
      return unreachable(variant)
  }
}
