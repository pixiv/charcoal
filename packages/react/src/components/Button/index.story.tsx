import { action } from '@storybook/addon-actions'
import React, { useRef } from 'react'
import { Story } from '../../_lib/compat'
import { ClickableElement } from '../Clickable'
import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    variant: {
      control: {
        type: 'inline-radio',
        options: ['Primary', 'Default', 'Overlay', 'Danger', 'Navigation'],
      },
    },
    size: {
      control: {
        type: 'inline-radio',
        options: ['S', 'M'],
      },
    },
  },
}

const DefaultStory = (args: ButtonProps) => (
  <Button {...args} onClick={action('click')}>
    Button
  </Button>
)

export const Default: Story<ButtonProps> = DefaultStory.bind({})

export const Primary: Story<ButtonProps> = DefaultStory.bind({})
Primary.args = {
  variant: 'Primary',
}

export const Navigation: Story<ButtonProps> = DefaultStory.bind({})
Navigation.args = {
  variant: 'Navigation',
}

export const Overlay: Story<ButtonProps> = DefaultStory.bind({})
Overlay.args = {
  variant: 'Overlay',
}

export const Danger: Story<ButtonProps> = DefaultStory.bind({})
Danger.args = {
  variant: 'Danger',
}

export const Small: Story<ButtonProps> = DefaultStory.bind({})
Small.args = {
  size: 'S',
}

export const Fixed: Story<ButtonProps> = DefaultStory.bind({})
Fixed.args = {
  fullWidth: true,
}

export const Disabled: Story<ButtonProps> = DefaultStory.bind({})
Disabled.args = {
  disabled: true,
}

export const Link: Story<ButtonProps<'a'>> = ({
  as,
  ...props
}: ButtonProps<'a'>) => {
  return (
    <Button {...props} as="a" href="#">
      href=#
    </Button>
  )
}

const NihongoStory = (args: ButtonProps) => (
  <Button {...args} onClick={action('click')}>
    日本語だよ
  </Button>
)

export const Nihongo: Story<ButtonProps> = NihongoStory.bind({})

const FocusStory = (args: ButtonProps) => <FocusStoryInternal {...args} />
const FocusStoryInternal = (args: ButtonProps) => {
  const ref = useRef<ClickableElement>(null)
  const focus = () => ref.current?.focus()
  return (
    <Button {...args} onMouseOver={focus} ref={ref}>
      Mouse over to focus
    </Button>
  )
}

export const Focus: Story<ButtonProps> = FocusStory.bind({})

const LayoutExampleStory = (args: ButtonProps) => (
  <div
    css={`
      display: grid;
      gap: 24px;
    `}
  >
    <div
      css={`
        display: grid;
        grid: auto / auto-flow min-content;
        gap: 8px;
      `}
    >
      <Button {...args}>Grid</Button>
      <Button {...args} variant="Navigation">
        Layout
      </Button>
      <Button {...args} variant="Danger">
        Sample
      </Button>
    </div>
    <div
      css={`
        display: grid;
        grid: auto-flow auto / 392px;
        gap: 8px;
      `}
    >
      <Button {...args} variant="Primary" fullWidth>
        Submit
      </Button>
      <Button {...args} variant="Default" fullWidth>
        Cancel
      </Button>
    </div>
    <div
      css={`
        display: flex;

        & > * + * {
          margin-left: 8px;
        }
      `}
    >
      <Button {...args}>Flex</Button>
      <Button {...args} variant="Navigation">
        Layout
      </Button>
      <Button {...args} variant="Danger">
        Sample
      </Button>
    </div>
    <div
      css={`
        display: flex;
        flex-direction: column;
        width: 392px;

        & > * + * {
          margin-top: 8px;
        }
      `}
    >
      <Button {...args} variant="Primary" fullWidth>
        Submit
      </Button>
      <Button {...args} variant="Default" fullWidth>
        Cancel
      </Button>
    </div>
    <div
      css={`
        display: flex;
        flex-direction: column;
        width: 392px;

        & > * + * {
          margin-top: 8px;
        }
      `}
    >
      <Button {...args} variant="Primary" fullWidth>
        すべて見る
      </Button>
      <Button {...args} variant="Default">
        作品を投稿
      </Button>
    </div>
  </div>
)

export const LayoutExample = LayoutExampleStory.bind({})
