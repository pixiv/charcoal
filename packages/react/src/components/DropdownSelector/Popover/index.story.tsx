import React, { useRef, CSSProperties, useState } from 'react'
import { Story } from '../../../_lib/compat'
import Popover, { PopoverProps } from '.'
import Button from '../../Button'

export default {
  title: 'DropdownSelector/Popover',
  component: Popover,
}

function Base(props: { style?: CSSProperties }) {
  const [isOpen, setIsOpen] = useState(false)
  const triggerRef = useRef(null)
  return (
    <>
      <Button
        onClick={() => {
          setIsOpen(true)
        }}
        style={props.style}
        ref={triggerRef}
      >
        button
      </Button>
      <Popover
        isOpen={isOpen}
        onClose={() => setIsOpen(false)}
        triggerRef={triggerRef}
      >
        <div style={{ margin: '8px 16px' }}>Hello</div>
      </Popover>
    </>
  )
}

export const Basic: Story<PopoverProps> = () => {
  return (
    <>
      <Base
        style={{
          position: 'absolute',
        }}
      />
      <Base
        style={{
          position: 'absolute',
          right: 8,
        }}
      />
      <Base
        style={{
          position: 'absolute',
          bottom: 8,
        }}
      />
      <Base
        style={{
          position: 'absolute',
          right: 8,
          bottom: 8,
        }}
      />
    </>
  )
}
