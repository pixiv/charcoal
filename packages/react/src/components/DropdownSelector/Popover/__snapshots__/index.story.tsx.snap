// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots DropdownSelector/Popover Basic 1`] = `
.c0 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c1 {
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  border-radius: 999999px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  padding-right: 24px;
  padding-left: 24px;
  color: var(--charcoal-text2);
  background-color: var(--charcoal-surface3);
  -webkit-transition: 0.2s color,0.2s background-color,0.2s box-shadow;
  transition: 0.2s color,0.2s background-color,0.2s box-shadow;
  height: 40px;
}

.c1:not(:disabled):not([aria-disabled]):active,
.c1[aria-disabled='false']:active,
.c1:not(:disabled):not([aria-disabled]):focus,
.c1[aria-disabled='false']:focus,
.c1:not(:disabled):not([aria-disabled]):focus-visible,
.c1[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1:not(:disabled):not([aria-disabled]):hover,
.c1[aria-disabled='false']:hover {
  color: var(--charcoal-text2-hover);
  background-color: var(--charcoal-surface3-hover);
}

.c1:not(:disabled):not([aria-disabled]):active,
.c1[aria-disabled='false']:active {
  color: var(--charcoal-text2-press);
  background-color: var(--charcoal-surface3-press);
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
}

<div
  data-dark={false}
>
  <button
    className="c0 c1"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "position": "absolute",
      }
    }
  >
    button
  </button>
  <button
    className="c0 c1"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "position": "absolute",
        "right": 8,
      }
    }
  >
    button
  </button>
  <button
    className="c0 c1"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "bottom": 8,
        "position": "absolute",
      }
    }
  >
    button
  </button>
  <button
    className="c0 c1"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "bottom": 8,
        "position": "absolute",
        "right": 8,
      }
    }
  >
    button
  </button>
</div>
`;
